File: lab2/KotlinArray/lexical-analyzer/LexicalAnalyzer.cpp
#include <iostream>
#include "LexicalAnalyzer.h"


LexicalAnalyzer::LexicalAnalyzer(std::string inputString) noexcept
    : inputString(std::move(inputString))
    , currentPosition(0)
    , currentChar('\0')
    , currentToken(Token::END)
{
    nextChar();
}


Token LexicalAnalyzer::getToken() noexcept {
    return currentToken;
}

std::size_t LexicalAnalyzer::getPosition() noexcept {
    return currentPosition;
}

void LexicalAnalyzer::nextToken() noexcept(false) {
    std::string builder;

    while (isSpace(currentChar)) {
        nextChar();
    }

    while (currentChar != '\0'
           && !isSpace(currentChar)
           && SpecialSymbols.find(currentChar) == std::string::npos) {

        builder += currentChar;
        nextChar();
    }

    if (Keywords.find(builder) != Keywords.end()) {
        currentToken = Token::KEYWORD;
    } else if (builder == Array) {
        currentToken = Token::ARRAY;
    } else if (!builder.empty()) {
        currentToken = Token::NAME;
    } else if (currentChar == ':') {
        nextChar();
        currentToken = Token::COLON;
    } else if (currentChar == '<') {
        nextChar();
        currentToken = Token::LANGLE;
    } else if (currentChar == '>') {
        nextChar();
        currentToken = Token::RANGLE;
    } else if (currentChar == ';') {
        nextChar();
        currentToken = Token::SEMICOLON;
    } else if (currentChar == '\0') {
        currentToken = Token::END;
    } else {
        std::string what = "Found unexpected symbol '";
        what += currentChar;
        what += "' at position ";
        what += std::to_string(currentPosition);
        throw ParseException(what);
    }
}


bool LexicalAnalyzer::isSpace(char ch) noexcept {
    return std::isspace(static_cast<int>(ch));
}

void LexicalAnalyzer::nextChar() noexcept {
    currentChar = currentPosition != inputString.size() ? inputString[currentPosition++] : '\0';
}

File: lab4/TranslatorGenerator/src/main/java/lexicalAnalyzer/LexicalAnalyzer.java
package lexicalAnalyzer;

import grammar.Terminal;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class LexicalAnalyzer {
    public final static Terminal END = new Terminal("<$>", false);
    private final Set<Terminal> terminals;
    private final String inputString;
    private int currentPosition;
    private char currentChar;
    private Terminal currentToken;


    public LexicalAnalyzer(final Set<Terminal> terminals, final String inputString) {
        this.terminals = terminals;
        this.inputString = inputString;
        this.currentPosition = 0;
        this.currentChar = '\0';
        this.currentToken = END;
        nextChar();
    }

    public Terminal getToken() {
        return currentToken;
    }

    public int getPosition() {
        return currentPosition - 1;
    }

    public void nextToken() throws ParseException {
//        System.err.printf("'currentPosition: %d, currentChar: %s%n", currentPosition, currentChar);
        StringBuilder stringBuilder = new StringBuilder();

        while (isSpace(currentChar)) {
            nextChar();
        }

//        System.err.printf("currentPosition: %d, currentChar: %s%n", currentPosition, currentChar);
        List<Terminal> previousMatches = new ArrayList<>();
        while (currentChar != '\0' && !isSpace(currentChar)) {
            stringBuilder.append(currentChar);
            nextChar();

            int matchesPrefixes = 0;
            List<Terminal> currentMatches = new ArrayList<>();
            for (Terminal terminal : terminals) {
                if (terminal.isRegexp()) {
                    if (stringBuilder.toString().matches(terminal.getValue())) {
                        ++matchesPrefixes;
                        currentMatches.add(new Terminal(terminal, stringBuilder.toString()));
                    }
                } else {
                    if (terminal.getValue().startsWith(stringBuilder.toString())) {
                        ++matchesPrefixes;
                    }
                    if (terminal.getValue().equals(stringBuilder.toString())) {
                        currentMatches.add(terminal);
                    }
                }
            }

            if (currentMatches.isEmpty() && matchesPrefixes == 0) {
                break;
            }

            previousMatches.addAll(currentMatches);
        }
//        System.err.printf("CurrentChar: %s, current position: %d, sb: %s%n", currentChar, currentPosition, stringBuilder);
//        System.err.println(previousMatches);

        if (previousMatches.isEmpty()) {
            if (currentPosition != inputString.length() + 1) {
                throw new ParseException(String.format("Found unexpected symbol '%s' at position %d", currentChar, currentPosition));
            }
            currentToken = END;
            return;
        }

        Terminal dummy = new Terminal(previousMatches.get(0), "");
        Terminal greedy = dummy;
        for (Terminal terminal : previousMatches) {
            if (terminal.getParsedValue().length() > greedy.getParsedValue().length() && !terminal.isRegexp()) {
                greedy = terminal;
            }
        }
        if (greedy == dummy) {
            for (Terminal terminal : previousMatches) {
                if (terminal.getParsedValue().length() > greedy.getParsedValue().length()) {
                    greedy = terminal;
                }
            }
        }

        for (int i = 0; i < stringBuilder.length() - greedy.getParsedValue().length(); ++i) {
            rollback();
        }

        currentToken = greedy;
    }

    public static void checkLexicalAnalyzer(final Set<Terminal> terminals, final String string) throws ParseException {
        LexicalAnalyzer lexicalAnalyzer = new LexicalAnalyzer(terminals, string);
        while (true) {
            lexicalAnalyzer.nextToken();
            Terminal token = lexicalAnalyzer.getToken();
            System.out.printf("%s ", token);
            if (token == END) {
                break;
            }
        }
        System.out.println();
    }


    private boolean isSpace(char ch) {
        return Character.isWhitespace(ch);
    }

    private void nextChar() {
        if (currentPosition != inputString.length()) {
            currentChar = inputString.charAt(currentPosition);
        } else {
            currentChar = '\0';
        }
        currentPosition++;
    }

    private void rollback() {
        assert(currentPosition != 0);
        if (currentPosition == 1) {
            currentChar = '\0';
            currentPosition = 0;
            return;
        }
        currentChar = inputString.charAt(currentPosition - 2);
        --currentPosition;
    }
}
